### PR TITLE
[WebAssembly][Object]Use file offset as function symbol address for linked files

### PR DESCRIPTION
**Pull request by dschuff**
WebAssembly doesn't have a single virtual memory space the way other object formats or architectures do, so "addresses" mean different things depending on the context.
Function symbol addresses in object files are offsets from the start of the code section. This is good for linking and relocation. However when dealing with linked binaries, offsets from the start of the file/module are more often used (e.g. for stack traces in browsers), and are more useful for use cases like binary size attribution. This PR changes Object to use the file offset instead of the section offset for function symbols, but only for linked (non-DSO) files.

This implements item number 4 from #76107

### CODE REVIEW

**Review by sbc100 — COMMENTED** [review]
What would happen if we instead just changes the offset to be absolute in all cases?

**Comment by dschuff** [comment]
@sbc100 we could potentially also do this for DSOs, with the reasoning that the use case is similar (e.g. such addresses could appear in stack traces, and users might want to do binary size attribution on DSOs as well). But I'm not sure if that would make linking worse (maybe not, since the code is PIC anyway, and the dynamic linker doesn't need to know code offsets?)

**Comment by dschuff** [comment]
Sorry, we raced.
If we did this for relocatable objects, I think the linker relies on this address being a section offset? I haven't tried it though.

**Comment by sbc100** [comment]
> Sorry, we raced. If we did this for relocatable objects, I think the linker relies on this address being a section offset? I haven't tried it though.

I yes, I suppose it must.   I wonder it would be simpler to update that.. of course that can be a followup too

**Comment by dschuff** [comment]
Generally speaking I think it's better for most purposes to have things be section-relative rather than file-relative anyway. You don't have to worry about changes to the known sections or custom sections like the dylink section that come before the code section affecting anything else.

File: llvm/lib/Object/WasmObjectFile.cpp [hunk 1]
```diff
@@ -1662,10 +1662,16 @@ Expected<StringRef> WasmObjectFile::getSymbolName(DataRefImpl Symb) const {
 Expected<uint64_t> WasmObjectFile::getSymbolAddress(DataRefImpl Symb) const {
   auto &Sym = getWasmSymbol(Symb);
   if (Sym.Info.Kind == wasm::WASM_SYMBOL_TYPE_FUNCTION &&
-      isDefinedFunctionIndex(Sym.Info.ElementIndex))
-    return getDefinedFunction(Sym.Info.ElementIndex).CodeSectionOffset;
-  else
-    return getSymbolValue(Symb);
+      isDefinedFunctionIndex(Sym.Info.ElementIndex)) {
+    // For object files, use the section offset. For linked files, use the file
+    // offset
```
Comment (sbc100) [inline llvm/lib/Object/WasmObjectFile.cpp:1667]: Perhaps mention here that file offset are using in other places such as backtraces and crash dumps?

**Review by sbc100 — APPROVED** [review]

**Comment by vitalybuka** [comment]
Breaks https://lab.llvm.org/buildbot/#/builders/5/builds/39751

**Comment by dschuff** [comment]
Thanks, working on a local repro. If I can't fix soon I'll revert.

**Pull request merged**